Multithreaded Inference Support

anirudh2290 opened the pull request:
## Description ##

To get some background on the PR, please go through the related github issue : https://github.com/apache/incubator-mxnet/issues/16431 . Below, I have added a brief overview for the PR.

## Brief Overview ##

1. This PR adds a multithreaded inference interface (EXPERIMENTAL) for MXNet. The approach it takes for this is to add a thread safe version of cached op with minimal set of features required for inference.
2. Adds C API to create cached op thread safe version: The API: CreateCachedOpEX takes an additional parameter `thread_safe`. When `thread_safe` is set to true, invokes on the cached op handle will invoke the thread safe version of cached op.
```
  /*!
   * \brief create cached operator, allows to choose thread_safe version
   * of cachedop
   */
  MXNET_DLL int MXCreateCachedOpEX(SymbolHandle handle,
                                   int num_flags,
                                   const char** keys,
                                   const char** vals,
                                   CachedOpHandle *out,
                                   bool thread_safe DEFAULT(false));
```

3. This PR adds tests and verification for thread safety of engine, for running parallel inference with the thread safe version of CachedOp with different models and different number of threads and different number of inferences per thread.
4. Model Testing has been limited to : resnet18, resnet50_v1, resnet152_v1
5. Adds CI stages and CI tests for running the thread safety tests. Please see : `capi-cpp-package GPU` in `unix-gpu` stage.

## Feedback/Acknowledgements ##

Below is the feedback/help received from offline code/design reviews:

1. @nswamy - Early inputs and offline design discussions related to this project, specifically with respect to dmlc::ThreadLocalStore.
2. @frankfliu - Explaining the requirements and use case for MXNet backend from a customer perspective, code review suggestions to have thread safe API for Create only and to determine whether to invoke cached op/cached op threadsafe based on handle.
3. @andreaolgiati  - Suggestion to document current threading model of MXNet (still a TODO), requesting adding support for dynamic shapes for NLP case (not supported yet), suggested documenting the long term plan with respect to thread safe inference in mxnet. (Documented in limitations and Future work section of the doc, will add required github issues).
4. @lanking520 , @larroy , @zachgk - Suggestion to add a new API for Threadsafe cached op version, since modifying existing one will break backward compatibility.
5. @frankfliu and @oorqueda - Suggestion for continuous maintenance (addition of support) of the models supported table. Will be maintained on the doc and RFC issue.
6. @samskalicky - offline code review suggestions with respect to code comments for locking strategy and comments for changes made for thread safe cached op.
7. @vikas-kum - better document the performance benefits. [still a TODO]

Thanks to all above for the suggestions, review and feedback.

Thanks to @marcoabreu and @arcadiaphy for the valuable comments on the RFC. 

Additionally, thanks to @roywei, @access2rohit , @stu1130 , @ChaiBapchya , @rondogency , @yuxihu, @vrakesh, @Keerthanvasist, @josephevans for code/design review.


## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented:
- For user-facing API changes, API doc string has been updated.
- For new C++ functions in header files, their functionalities and arguments are documented.
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

zachgk commented on include/mxnet/c_api.h:
```diff
@@ -1274,10 +1274,28 @@ MXNET_DLL int MXCreateCachedOpEx(SymbolHandle handle,
+                                 const char** keys,
+                                 const char** vals,
+                                 CachedOpHandle *out,
+                                 bool thread_safe = false);
```
Aren't default arguments like this only supported in C++ and this file is just C?

anirudh2290 replied:
Yes i need to DEFAULT(false) like https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/c_api.h#L3157

larroy commented on ci/docker/runtime_functions.sh:
```diff
@@ -1353,6 +1373,19 @@ integrationtest_ubuntu_gpu_cpp_package() {
+    python3 -c "import mxnet as mx; mx.test_utils.download_model(\"imagenet1k-resnet-18\"); mx.test_utils.download_model(\"imagenet1k-resnet-152\"); mx.test_utils.download_model(\"imagenet1k-resnet-50\");"
+    build/tests/cpp/mxnet_unit_tests --gtest_filter="ThreadSafety.*"
+    build/tests/cpp/mxnet_unit_tests --gtest_filter="ThreadSafety.*" --thread-safety-with-cpu
+    export MXNET_ENGINE_TYPE=NaiveEngine
```
Why NaiveEngine here? Could you add comments?

anirudh2290 replied:
This is to run thread safety tests in NaiveEngine mode. I have added a comment.

anirudh2290 commented:
@zachgk @larroy Thanks for the review. I have addressed it.

larroy commented on ci/docker/runtime_functions.sh:
```diff
@@ -1212,6 +1230,8 @@ unittest_ubuntu_cpugpu_perl() {
 unittest_cpp() {
     set -ex
+    export PYTHONPATH=./python/
+    python3 -c "import mxnet as mx; mx.test_utils.download_model(\"imagenet1k-resnet-18\"); mx.test_utils.download_model(\"imagenet1k-resnet-152\"); mx.test_utils.download_model(\"imagenet1k-resnet-50\");"
```
I don't like this inlined python very much, it's also duplicated code below, it could be factored out into a utility script or a function.

anirudh2290 replied:
I think writing a function for this is an overkill.

larroy commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
````diff
@@ -0,0 +1,324 @@
+To download a model use the `get_model.py` script. This downloads a model to run inference.
+
+```python
+python3 get_model.py --model <model_name>
````
why not use this get model above?

larroy commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
```diff
@@ -0,0 +1,324 @@
+    return EXIT_FAILURE;
+  }
+  std::string model_name = std::string(argv[1]);
+  int num_threads = std::atoi(argv[2]);
```
let's use std::to_string instead of unsafe atoi in general. This should be a linter rule.

anirudh2290 replied:
If you are hinting at the undefined behavior that will happen only for outside int ranges. As this is for num_threads for an example script this is unlikely. Otherwise, I think atoi seems fine for the use case here.

larroy commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
```diff
@@ -0,0 +1,324 @@
+  std::map<std::string, mxnet::cpp::NDArray> parameters;
+  mxnet::cpp::NDArray::Load(param_file, 0, &parameters);
+
+  for (std::string name : out.ListInputs()) {
```
reference?  does this code compile? Can be included in the doc? how do we keep it current.

anirudh2290 replied:
no, this is a doc and I had added this to refer to the code that I was refering to in the cc file. I changed it to the code blocks in the cc file, to make it easier to maintain.

larroy commented on example/multi_threaded_inference/multi_threaded_inference.cc:
```diff
@@ -0,0 +1,351 @@
+  std::map<std::string, mxnet::cpp::NDArray> parameters;
+  mxnet::cpp::NDArray::Load(param_file, 0, &parameters);
+
+  for (std::string name : out.ListInputs()) {
```
const reference?

anirudh2290 replied:
done.

larroy commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,180 @@
+#include "./imperative_utils.h"
 
 namespace mxnet {
+namespace {
```
does this anon ns need to be inside mxnet?

anirudh2290 replied:
this was from before. it was to keep these functions local to the translation unit.

larroy replied:
Anon namespace is fine, the question is why is inside mxnet when is dealing with nnvm.

anirudh2290 replied:
This is specific to cached op and mxnet. The location is correct.

larroy commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,180 @@
+                      const std::string& s) {
+  return prefix + "_" + s;
+}
+void CreateFullGraph(const nnvm::Symbol& sym,
```
missing function documentation

anirudh2290 replied:
this function was from before. added doc.

larroy commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,180 @@
+    }
+  }
+
+  bool do_elim_common_expr = dmlc::GetEnv("MXNET_ELIMINATE_COMMON_EXPR", true);
```
can this be factored out in smaller functions?

anirudh2290 replied:
this function was from before, its outside the scope of this pr. Feel free to add a new PR for refactoring this.

larroy commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,180 @@
+  }
+}
+
+void SetRefCounts(nnvm::Graph* fwd_graph, const nnvm::Graph& full_graph) {
```
doc?

anirudh2290 replied:
added doc.

larroy commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,180 @@
+void OptimizeGraph(nnvm::Graph * full_graph, nnvm::Graph * fwd_graph, nnvm::Graph * grad_graph,
+                   const Context& context, size_t num_forward_outputs, const bool inlining) {
+#if MXNET_USE_CUDA && !defined(_WIN32)
+  if (context.dev_mask() == kGPU &&
```
can this be factored out?

anirudh2290 replied:
this function was from before, refactoring this is outside the scope of this pr.

larroy commented on src/imperative/cached_op_threadsafe.cc:
```diff
@@ -0,0 +1,373 @@
+      CHECK(config_.static_alloc) << "static_alloc must be True when static_shape is True";
+  }
+
+  // construct forward graph
```
why not factor out instead of putting a comment and a block of code?

anirudh2290 replied:
refactored.

larroy commented on src/imperative/cached_op_threadsafe.cc:
```diff
@@ -0,0 +1,373 @@
+  }
+}
+
+OpStatePtr CachedOpThreadSafe::DynamicForward(const Context& default_ctx,
```
doc?

anirudh2290 replied:
done.

larroy commented on src/imperative/cached_op_threadsafe.cc:
```diff
@@ -0,0 +1,373 @@
+  // Allocate entries
+  // This buff is thread local and used to store intermediate
+  // nodes in the graph
+  buff.resize(idx.num_node_entries());
```
This coding style is similar as in the backward pass, and I'm not a fan of this code style because I find it not very readable due to back and forth the indexed graph, indices, buffers and arrays. I think it can be abstracted more and make it easier to follow by factoring out in functions where the indexes don't appear. In the end it's the way a big chunk of mxnet is so this is like another drop in the ocean, but anyway.

anirudh2290 replied:
feel free to open a new pr for refactor.

larroy replied:
You are changing this code now, there's should be no need to immediately refactor in the future.

anirudh2290 replied:
i am just following existing style. Refactoring cached op is not in the scope of this PR.

eric-haibin-lin commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
```diff
@@ -0,0 +1,199 @@
+Examining the current state of thread safety in MXNet we can arrive to the following conclusion:
+
+1. MXNet Dependency Engine is thread safe (except for WaitToRead invoked inside a spawned thread. Please see Limitations section)
+2. Graph Executor which is Module/Symbolic/C Predict API backend is not thread safe
```
Awesome documetnation and knowledge sharing! Could you share what class/component used in graph executor/CachedOp are not thread safe?

anirudh2290 replied:
Thanks for the kind words @eric-haibin-lin . For graph executor,  I saw core dumps with memory leaks for graph executor bind which indicated that Graph Executor Bind was not thread safe. I didnt dig more into what exactly in Graph Executor bind, it could be one of the NNVM passes. For cached op, the issues where 1. with respect to not using thread local variables for intermediate states buffer (buff) in CachedOp::DynamicForward 2. Issues related to hang in accept4 call in cuda lib when  op push is not serialized in CachedOp::Forward.

apeforest commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
````diff
@@ -0,0 +1,199 @@
+```python
+python3 get_model.py --model imagenet1k-inception-bn
+```
+Only the supported models with `get_model.py` work with multi threaded inference.
````
Where can the user find the supported model names? Can you provide a link here?

apeforest commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
```diff
@@ -0,0 +1,199 @@
+- Build the multi-threaded inference example
+
+### Setup the MXNet C++ API
+To use the C++ API in MXNet, you need to build MXNet from source with C++ package. Please follow the [built from source guide](/get_started/ubuntu_setup.html), and [C++ Package documentation](/api/cpp)
```
Can I just use C API without building the cpp package?

anirudh2290 replied:
I used CPP package because using it makes writing these tests and example a little easier. It can also be written solely with C API.

apeforest commented on example/multi_threaded_inference/multi_threaded_inference.cc:
```diff
@@ -0,0 +1,351 @@
+#include <mxnet/c_predict_api.h>
+#include "mxnet-cpp/MxNetCpp.h"
+
+const mx_float DEFAULT_MEAN = 117.0;
```
Let's just use float. I don't think we will need this typedef any more.

anirudh2290 replied:
changed to float

apeforest commented on example/multi_threaded_inference/multi_threaded_inference.cc:
```diff
@@ -0,0 +1,351 @@
+}
+
+// Run inference on a model
+void run_inference(const std::string& model_name, const std::vector<mxnet::cpp::NDArray>& input_arrs,
```
just mxnet::NDArray?

anirudh2290 replied:
the cpp ndarray is a frontend ndarray which is implemented like any other frontend ndarray i.e. save the ndarray handle (pointer to the ndarray) and perform C API operations on it.

apeforest commented on example/multi_threaded_inference/multi_threaded_inference.cc:
```diff
@@ -0,0 +1,351 @@
+
+  // Prepare context
+# if MXNET_USE_CUDA == 1
+  mxnet::Context backend_ctx;
```
Same question for mxnet::Context vs mxnet::cpp::Context

anirudh2290 replied:
frontend version of the context data structure. provides getters and setters for dev_type and dev_id.

apeforest commented on example/multi_threaded_inference/multi_threaded_inference.cc:
```diff
@@ -0,0 +1,351 @@
+  CHECK(num_threads == argc - 4) << "Number of files provided, should be same as num_threads";
+  std::vector<std::string> test_files;
+  for (size_t i = 0; i < argc - 4; ++i) {
+  test_files.emplace_back(argv[4 + i]);
```
indentation

anirudh2290 replied:
fixed.

apeforest commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,195 @@
+void CreateForwardGraph(const nnvm::Symbol &sym, nnvm::Graph *fwd_graph) {
+  using namespace nnvm;
+  static const auto _copy_op = Op::Get("_copy");
+  {
```
Can we write a function instead of using braces to mark the scope.

anirudh2290 replied:
this was part of a refactor in this pr. earlier this whole chunk of code was called directly from forward of cached op. The braces are not needed here. I have removed it.

apeforest commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,195 @@
+    *fwd_graph = exec::EliminateCommonExpr(std::move(*fwd_graph));
+
+  // construct backward graph
+  {
```
Same here. Using a function instead of a pair of braces can improve readability.

anirudh2290 replied:
this is the only place this function is called from. I think we can wait till this is needed elsewhere and move it to its own function.

larroy replied:
Is not about being used elsewhere, is about readability. If you don't need to dive deep is better to read a function call and skip the details.

anirudh2290 replied:
Added a new function.

apeforest commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,195 @@
+  }
+
+  // construct full graph
+  {
```
Same here.

anirudh2290 replied:
similar to above, i think we can wait till this is needed elsewhere and move it to its own function.

apeforest commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,195 @@
+                                                     full_graph->outputs.end());
+  SetRefCounts(fwd_graph, *full_graph);
+}
+
```
remove redundant blank lines.

anirudh2290 replied:
removed.

apeforest commented on src/imperative/cached_op.cc:
```diff
@@ -32,247 +32,12 @@ DMLC_REGISTER_PARAMETER(CachedOpConfig);
   std::vector<OpStatePtr> op_states;
 };
 
-void CreateFullGraph(const nnvm::Symbol& sym,
```
Any particular reason you moved these function implementation from .cc to .h?

anirudh2290 replied:
to allow to be used in both CachedOp and CachedOpThreadSafe.

apeforest commented on src/imperative/cached_op_threadsafe.h:
```diff
@@ -0,0 +1,129 @@
+
+
+
+class CachedOpThreadSafe : public CachedOp {
```
Maybe I missed something, but why can't we modify the CachedOp to make it thread safe rather than creating a new inherited class? In which scenarios users want to choose a thread unsafe CachedOp and is there any guideline somewhere?

anirudh2290 replied:
the thread safe cached op has not achieved feature parity with the current cached op. Lot of things not supported including training, bulking, multi context use case etc.  Guideline is to use the thread safe cached op for multi threaded inference use case and default one for everything else.

apeforest commented on tests/cpp/operator/mkldnn_operator_test.cc:
```diff
@@ -719,7 +707,10 @@ void TestOpEx(const OpAttrs &forward_attrs, const OpAttrs &backwards_attrs) {
             DispatchMode::kFComputeEx, mxnet::OpStatePtr());
         Engine::Get()->WaitForAll();
-        AssertEqual(outputs, ex_outputs);
+        // TODO(unassigned): Need to fix op, should work for the whole vector
```
Can you assign TODO here? otherwise I am afraid it might remain TODO forever :)

anirudh2290 replied:
i didn't know who to assign it to. I have assigned to @pengzhao-intel for now, please feel free to suggest other names.

apeforest commented:
Thanks for this big contribution! In the future, it might be easier for reviewers if the PR can be broken into smaller pieces :)

ChaiBapchya commented on src/imperative/cached_op_threadsafe.cc:
```diff
@@ -0,0 +1,342 @@
+  SetForwardRefCounts(&fwd_graph_);
+
+  // Set param indices
+  {
```
Same as pointed by Lin and Pedro before. Factor out to a function.

ChaiBapchya commented on tests/cpp/engine/thread_local_test.cc:
```diff
@@ -56,7 +56,7 @@ static int ThreadSafetyTest(int num, std::vector<int>* tmp_inputs, std::vector<i
     return 0;
 }
 
-TEST(ThreadLocal, verify_thread_safety) {
```
Why is this changed?

anirudh2290 replied:
just making the test naming consistent.

ChaiBapchya commented on tests/cpp/thread_safety/thread_safety_test.cc:
```diff
@@ -0,0 +1,667 @@
+                                  num_threads, &result_expected, &hdl);
+
+
+    // Create thread safe cahced op
```
```suggestion
    // Create thread safe cached op
```

eric-haibin-lin commented on src/imperative/cached_op.cc:
```diff
@@ -32,247 +32,12 @@ DMLC_REGISTER_PARAMETER(CachedOpConfig);
 
 constexpr uint32_t kEidNotExist = std::numeric_limits<uint32_t>::max();
 
-const char CachedOp::FULL[] = "full";
```
Could you describe the main changes made for cached_op.cc/h (code moved/removed/added)?

anirudh2290 replied:
added comments to cached_op.cc, moved the constant strings and the helper functions AddPrefix, CreateFullGraph, SetRefCounts, OptimizeRefCounts, CachedOpState to cached_op.h file. Refactored CreateForwardGraph and SetForwardRefCounts into its own function.

apeforest replied:
Could you add these to PR description?

anirudh2290 commented on CMakeLists.txt:
```diff
@@ -825,7 +829,6 @@ if(MSVC AND USE_MXNET_LIB_NAMING)
   set_target_properties(mxnet PROPERTIES OUTPUT_NAME "libmxnet")
 endif()
 
-add_subdirectory(tests)
```
this is for the tests to be built after the cpp package, because of the additional dependency on cpp package for some tests.

anirudh2290 commented on cpp-package/include/mxnet-cpp/symbol.hpp:
```diff
@@ -151,6 +151,18 @@ inline std::vector<std::string> Symbol::ListArguments() const {
+
+inline std::vector<std::string> Symbol::ListInputs() const {
+  std::vector<std::string> ret;
+  mx_uint size;
```
can you elaborate ?

anirudh2290 commented on example/multi_threaded_inference/Makefile:
```diff
@@ -0,0 +1,67 @@
+# If MXNET_LIB_DIR env variable set use that, otherwise defaults to MXNET_ROOT/build
+ifndef MXNET_LIB_DIR
+    MXNET_LIB_DIR=$(MXNET_ROOT)/lib
+    # cmake default by default
```
hehe, not sure either. will remove it!

anirudh2290 commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
```diff
@@ -0,0 +1,199 @@
+
+The above script spawns 2 threads, shares the same cachedop and params among two threads, and runs inference on GPU. It returns the inference results in the order in which files are provided.
+
+NOTE: This example is to demonstrate the multi-threaded-inference with cached op. The inference results work well only with specific models (e.g. imagenet1k-inception-bn). The results may not necessarily be very accurate because of different preprocessing step required etc.
```
What I mean is that if you use a random image to test here. the inference results may not be very accurate. the width and height of image is hardcoded and in the example and it may require specific preprocessing steps depending on image to get accurate result.

anirudh2290 commented on docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md:
```diff
@@ -0,0 +1,199 @@
+[https://github.com/apache/incubator-mxnet/example/multi_threaded_inference/multi_threaded_inference.cc#L248-L262](multi_threaded_inference.cc#L248-262)
+
+The above creates the lambda function taking the thread number as the argument.
+If `random_sleep` is set it will sleep for a random number (secs) generated between 0 to 5 seconds.
```
it can be 0-10 also, i just wanted to simulate a multi-threaded scenario where different threads call forward at different times.

anirudh2290 commented on example/multi_threaded_inference/multi_threaded_inference.cc:
```diff
@@ -0,0 +1,351 @@
+  std::string model_name = std::string(argv[1]);
+  int num_threads = std::atoi(argv[2]);
+  bool is_gpu = std::atoi(argv[3]);
+  CHECK(num_threads == argc - 4) << "Number of files provided, should be same as num_threads";
```
good point, i can change this.

larroy commented on src/imperative/cached_op.h:
```diff
@@ -26,8 +26,191 @@
+void CreateForwardGraph(const nnvm::Symbol &sym, nnvm::Graph *fwd_graph) {
+  using namespace nnvm;
+  static const auto _copy_op = Op::Get("_copy");
+  NodeEntryMap<size_t> dedup_out;
```
Why this needs to map to size_t when you are treating just as a set?

anirudh2290 replied:
its not a set, NodeEntryMap is typedef'd to following in NNVM:   

template<typename ValueType>
  using NodeEntryMap = std::unordered_map<NodeEntry, ValueType, NodeEntryHash, NodeEntryEqual>;

apeforest commented on src/imperative/cached_op_threadsafe.cc:
```diff
@@ -0,0 +1,329 @@
+  using namespace imperative;
+
+  {
+  auto state_ptr = GetCachedOpState(default_ctx);
```
indent? Also, better to refactor into a function.

apeforest commented on src/imperative/cached_op_threadsafe.cc:
```diff
@@ -0,0 +1,329 @@
+  auto state_ptr = GetCachedOpState(default_ctx);
+  auto op_state = OpStatePtr::Create<DynamicRuntime>();
+  auto &runtime = op_state.get_state<DynamicRuntime>();
+  {
```
Refactor into a function would be much better for readability

eric-haibin-lin commented on src/imperative/cached_op_threadsafe.cc:
```diff
@@ -0,0 +1,329 @@
+
+namespace mxnet {
+
+DMLC_REGISTER_PARAMETER(CachedOpThreadSafeConfig);
```
Is it necessary to have another implementation for cached op? What does it take to make the original cached op thread safe? This makes the code hard to maintain. Is there a plan of future works of thread_safe cached op? Do we plan to merge it back to original cached op?

rondogency replied:
anirudh answered here: https://github.com/apache/incubator-mxnet/pull/16654#discussion_r366074941 but I think if we can complete all the todo items it's better to merge back

anirudh2290 replied:
> Is it necessary to have another implementation for cached op? What does it take to make the original cached op thread safe? This makes the code hard to maintain.

Hi @eric-haibin-lin , the scope of this project was to allow for multi threaded inference with the cached op. This is used in the DJL project (https://github.com/awslabs/djl). There are two options for implementation : 1. either include all these changes with a different code path inside cached op code or 2. create a new implementation for thread safe cached op. Between the two, I find the latter to be easier to maintain. To make the existing cached op thread safe (as @rondogency pointed out), it would require training support, bulking support, multi context support, subgraph api support, dynamic shape support, ensuring that there are no perf regressions. At the very least, there is a lot of additional testing required. 

> Is there a plan of future works of thread_safe cached op?

For the future works of the thread_safe cached op, at this point we want to continue maintaining and addressing customer feedback for the current thread safe cached op version and address performance issues. 

> Do we plan to merge it back to original cached op?

I am not aware of a plan for this currently.

rondogency commented:
@anirudh2290 Thanks for this big contribution! one suggestion is that lots of developers complaining about cached op lacking docblocks and comments. since you're familiar with cached op, can you add a bit more docs on either your cached op or the original one? It is one time effort and can benefit the whole community

anirudh2290 commented:
@rondogency I am happy to add more comments for the thread safe cached op or the code that I have touched in this PR.

anirudh2290 commented:
> On the other hand, I am not sure if we can accept this cached op without bulking and subgraph support. Subgraph is used for multiple accelerators for MXNet inference. Also, bulking is essential to inference speed on GPUs. For instance, we saw 20% perf regression if bulking scope is smaller than the optimal scope #9055 and it would be even more significant if no bulking is enabled at all. I am not sure if this can be shipped with these unknown limitations. Thinking about how to better support thread safe bulking and subgraph may affect the implementation and design.

1. I have added bulking support. 

2. For subgraphing, If I am correct, MXNet still doesnt support subgraph API with gluon. It is still supported only with symbol and MXNET_SUBGRAPH_BACKEND env variable works only with graph executor. What I meant to say is that the param subgraph in CachedOp is not supported. Now you can use a symbol which has already been converted and replaced with subgraph ops with the cached_op_threadsafe version. I have added a test to demonstrate that. I think this should address your concern with respect to subgraphing.

> Do we have evidence that the newly introduced cached op is performant? I'm concerned about the mutex for the whole forward function - that means no thread can concurrently push operations to the engine. What I observed is - that could take a long time depending on the model used.

Although the mutex in forward is just pushing the ops, which may have delays between threads, the execution of these ops will be scheduled in parallel, especially when there is no dependency from the ops different threads and should improve performance.

https://github.com/awslabs/djl did some benchmarks and were able to see improved throughput upto 2x with naive engine. I was also able to see performance improvements upto 1.7x when you increase the number of worker threads on the threaded engine for cpu.

anirudh2290 commented:
I have addressed all the comments here. Please let me know if you have additional comments.

pengzhao-intel commented:
We're on the long vocation and will back to office in 10th Feb.

anirudh2290 commented:
Thanks a lot for your kind words Chris! Appreciate your dedication towards the mxnet community!

cjolivier01 commented:
hahaha...

On Thu, Jan 30, 2020 at 8:41 PM Anirudh Subramanian <
notifications@github.com> wrote:

> Thanks a lot for your kind words Chris! Appreciate your dedication towards
> the mxnet community!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-mxnet/pull/16654?email_source=notifications&email_token=ACVWZ7INFTQHSQ65ICMDIXTRAOTZBA5CNFSM4JFW2Z22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKNO4KY#issuecomment-580578859>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACVWZ7JKKZD7RO4IKJUZHFDRAOTZBANCNFSM4JFW2Z2Q>
> .
>

larroy commented:
Did you run performance benchmarks to see that there's no regression? as I understand this is a change that could have an impact on performance in the engine. Would be useful to run the usual models we have in the performance dashboards.

leezu commented:
@anirudh2290 this seems to have broken C++ examples. See https://github.com/apache/incubator-mxnet/issues/17514

eric-haibin-lin commented:
@zhreshold FYI
